Add support for final constants

kocsismate opened the pull request:
RFC: https://wiki.php.net/rfc/final_class_const

kamil-tekiela commented:
Why was this not supported before?

kocsismate commented:
> Why was this not supported before?

I honestly don't have an idea. Maybe it didn't seem important enough (?).

mvorisek commented:
https://3v4l.org/EYL34 - `the final modifier is allowed only for methods and classes`

That error should be updated as well.

reyadkhan commented:
Would be better if we could use the final keyword with method parameter like:  function addFoo(final Foo $foo): array {}

kocsismate commented:
> Would be better if we could use the final keyword with method parameter like: function addFoo(final Foo $foo): array {}

What's the use-case here?

reyadkhan commented:
> > Would be better if we could use the final keyword with method parameter like: function addFoo(final Foo $foo): array {}
> 
> What's the use-case here?

The parameter variable shouldn't re-assign inside the method.

drealecs commented:
> > > Would be better if we could use the final keyword with method parameter like: function addFoo(final Foo $foo): array {}
> > 
> > What's the use-case here?
> 
> The parameter variable can't be reassigned inside the method.
> It's frequently used in some programming language as a good practice like 'Java'

That use case comes from Java where variables and parameters are constants, once defined. In php the keyword `final` is strictly related to now allowing it changed in extending classes or not allowing extending at all.

I think the use case mentioned is valid but probably a new keyword would be better: const, readonly.

Also, it would be completely unrelated with this change.
But I believe it would be an important addition to the language as (mostly) all variables should not be reassigned. Will reduce the maintenance cost and bugs.

TysonAndre commented on Zend/tests/constants/final_constants/final_const3.phpt:
```diff
@@ -0,0 +1,13 @@
+
+?>
+--EXPECTF--
+Fatal error: Private constant Foo::A cannot be final as it is never overridden in %s on line %d
```
Not really a major issue since it's impossible for existing code to use `private final const` so there's no bc break.

It seems allowed to override private constants with public constants and static:: will use the public constants, e.g. in php 7.4

Also, the fact that `private final` is forbidden isn't documented in the rfc right now? https://wiki.php.net/rfc/inheritance_private_methods

```
php > class A { public const A = 'const A::A'; public static function dump() { var_dump(static::A); } }
php > class B extends A { public const A = 'const B::A'; }
php > B::dump();
string(10) "const B::A"


php > class X { private function foo() { echo "in X::foo\n";} public function dump() { $this->foo(); } }
php > class Y extends X { public function foo() { echo "In Y::foo\n"; } }
php > (new Y())->dump();
in X::foo
```

-------

I guess updating the issue message may be better, e.g. `as it would prevent subclasses from accessing a constant of that name` but I don't know why that would be a compelling reason to forbid `private final` - it'd also be a compile error if those subclasses tried to declare the constant to override `protected final`

Somewhat reminds me of https://externals.io/message/110540#110576

kocsismate replied:
First of all, thank you for the review!

> Also, the fact that private final is forbidden isn't documented in the rfc right now?

Yeah, this is quite unfortunate... I realized that it's not documented just after starting the vote. :(

> I don't know why that would be a compelling reason to forbid private final - it'd also be a compile error if those subclasses tried to declare the constant to override protected final

The reason why `final private` is forbidden is to be consistent with https://wiki.php.net/rfc/inheritance_private_methods indeed. So In my opinion (which might be wrong of course) if private methods (- constructors) cannot be final then the same should apply for constants. Unfortunately, I'm not 100% sure what exact issue is that you see, so could you please clarify it?

nikic commented on Zend/tests/constants/final_constants/final_const3.phpt:
```diff
@@ -0,0 +1,13 @@
+
+?>
+--EXPECTF--
+Fatal error: Private constant Foo::A cannot be final as it is never overridden in %s on line %d
```
```suggestion
Fatal error: Private constant Foo::A cannot be final as it is not visible to other classes in %s on line %d
```
or something like that? I found the message a bit confusing.

Or possibly don't try to include reasoning at all:
```suggestion
Fatal error: Private constant Foo::A cannot be final in %s on line %d
```

kocsismate replied:
In my opinion the including the reason really helps understand why the error happens. Since the message format is copy pasted from the similar one for methods, I'll adapt that one as well.

nikic commented on Zend/zend_API.c:
```diff
@@ -4316,7 +4316,7 @@ ZEND_API zend_class_constant *zend_declare_class_constant_ex(zend_class_entry *c
 	zend_class_constant *c;
 
 	if (ce->ce_flags & ZEND_ACC_INTERFACE) {
-		if (access_type != ZEND_ACC_PUBLIC) {
```
Maybe renamed access_type -> flags?

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -3699,6 +3699,12 @@ ZEND_METHOD(ReflectionClassConstant, isProtected)
+/* Returns whether this constant is final */
+ZEND_METHOD(ReflectionClassConstant, isFinal)
+{
+	_class_constant_check_flag(INTERNAL_FUNCTION_PARAM_PASSTHRU, ZEND_ACC_FINAL);
```
You also need to add it to the getModifiers() mask and add an IS_FINAL constant to the class.

nikic commented:
Ping :) Please tell me if you need help.

kocsismate commented:
> Ping :) Please tell me if you need help.

Thanks for the offer, I'll definitely ask for help :) TBH I don't really have much free time these days, so I haven't yet tried to fix the issues you pointed out. I'm going to take a stab this weekend, unless I completely get passed out by the 2nd vaccine ^^

kocsismate commented:
I addressed all the review comments, except for the most important one :( I just couldn't find out how to adjust the inheritance rules. Or should we track if a constant was inherited when adding them to the constant table (e.g. by using a new flag)? Or is it still not enough to do?

nikic commented:
@kocsismate Does this look right to you?

kocsismate commented:
> Does this look right to you?

Yeah, very nice! :)

kocsismate commented:
Closing via https://github.com/php/php-src/commit/a5360e80c27fcf2e1a3e4c4602940fd1d7a8ce89
